coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -269,45 +304,147 @@ class SingleClusterPlanner(val dataset: Dataset,
       case lp: ScalarFixedDoublePlan       => materializeFixedScalar(qContext, lp)
-      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp)
-      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp)
-      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
-      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
-      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp)
+      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp, forceInProcess)
+      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp, forceInProcess)
+      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp, forceInProcess)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp, forceInProcess)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+  /**
+   * Returns the shards spanned by a LogicalPlan's data.
+   * @return an occupied Option iff shards could be determined for all leaf-level plans.
+   */
+  private def getShardSpanFromLp(lp: LogicalPlan,
+                                 qContext: QueryContext): Option[Set[Int]] = {
+
+    case class LeafInfo(lp: LogicalPlan,
+                        filters: Set[ColumnFilter],
+                        tslabels: Option[Seq[String]])
+
+    // construct a LeafInfo for each leaf
+    val targetSchemaProvider = qContext.plannerParams.targetSchemaProvider
+    val leaves = findLeafLogicalPlans(lp).map{ leafPlan =>
+      val filters = getColumnFilterGroup(leafPlan)
+      assert(filters.size == 1, s"expected leaf plan to yield single filter group, but got ${filters.size}")
+      if (targetSchemaProvider.isDefined) {
+        // store the target schema labels (if they're defined) into the LeafInfo
+        val tsFunc = targetSchemaProvider.get.targetSchemaFunc(filters.head.toSeq)
+        val (start, stop) = leafPlan match {
+          // TODO(a_theimer): other leaf plan types? Range selector types?
+          case rs: RawSeries => rs.rangeSelector match {
+            case is: IntervalSelector => (is.from, is.to)
+            case _ => throw new IllegalArgumentException(s"unhandled type: ${rs.rangeSelector.getClass}")
```

Comment:
I see I've handled this differently than in `getUniversalTargetSchema`. I'll changes this to instead return None if the IntervalSelector is unrecognized. Maybe I should do this for the unrecognized LogicalPlan case, as well?